plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/ProgramConversionContext.kt

```diff
@@ -0,0 +1,181 @@
+import org.jetbrains.kotlin.fir.types.isInt
+import org.jetbrains.kotlin.fir.types.isUnit
+import org.jetbrains.kotlin.fir.visitors.FirVisitor
+import org.jetbrains.kotlin.formver.scala.emptySeq
+import org.jetbrains.kotlin.formver.scala.seqOf
+import org.jetbrains.kotlin.formver.scala.silicon.ast.*
+import org.jetbrains.kotlin.formver.scala.silicon.ast.Stmt.Seqn
+import org.jetbrains.kotlin.formver.scala.toScalaBigInt
+import org.jetbrains.kotlin.formver.scala.toScalaSeq
+import org.jetbrains.kotlin.text
+import org.jetbrains.kotlin.types.ConstantValueKind
+import viper.silver.ast.Declaration
+import viper.silver.ast.LocalVarDecl
+import viper.silver.ast.Method
+import viper.silver.ast.Program
+
+const val INT_BACKING_FIELD = "backing_int"
+const val RETURN_VARIABLE_NAME = "ret"
+
+class ConvertedVar(val name: String, val type: ConvertedNonUnitType) {
+    fun toLocalVarDecl(
+        pos: Position = Position.NoPosition,
+        info: Info = Info.NoInfo,
+        trafos: Trafos = Trafos.NoTrafos,
+    ): LocalVarDecl = localVarDecl(name, type.viperType, pos, info, trafos)
+
+    fun toLocalVar(
+        pos: Position = Position.NoPosition,
+        info: Info = Info.NoInfo,
+        trafos: Trafos = Trafos.NoTrafos,
+    ): Exp.LocalVar = Exp.LocalVar(name, type.viperType, pos, info, trafos)
+
+    fun preconditions(): List<Exp> = type.preconditions(toLocalVar())
+    fun postconditions(): List<Exp> = type.postconditions(toLocalVar())
+}
+
+class ConvertedMethodSignature(val name: String, val params: List<ConvertedVar>, val returns: List<ConvertedVar>) {
+    fun toMethod(
+        pres: List<Exp>, posts: List<Exp>,
+        body: Seqn?,
+        pos: Position = Position.NoPosition,
+        info: Info = Info.NoInfo,
+        trafos: Trafos = Trafos.NoTrafos,
+    ): Method =
+        method(
+            name,
+            params.map { it.toLocalVarDecl() },
+            returns.map { it.toLocalVarDecl() },
+            params.flatMap { it.preconditions() } + pres,
+            params.flatMap { it.postconditions() } +
+                    returns.flatMap { it.preconditions() } + posts,
+            body, pos, info, trafos,
+        )
+}
+
+class ProgramConversionContext {
+    private val methods: MutableList<Method> = mutableListOf()
+
+    val program: Program
+        get() = Program(
+            emptySeq(), /* Domains */
+            seqOf(field(INT_BACKING_FIELD, Type.Int)), /* Fields */
+            emptySeq(), /* Functions */
+            emptySeq(), /* Predicates */
+            methods.toScalaSeq(), /* Functions */
+            emptySeq(), /* Extensions */
+            Position.NoPosition.toViper(),
+            Info.NoInfo.toViper(),
+            Trafos.NoTrafos.toViper()
+        )
+
+    fun addWithBody(declaration: FirSimpleFunction) {
+        val methodCtx = MethodConversionContext(this, declaration);
+        methods.add(methodCtx.fullMethod)
+    }
+
+    fun convertType(type: ConeKotlinType): ConvertedType {
```

Comment:
I think you can rewrite this using a `when` expression. Like this:

```kotlin
fun convertType(type: ConeKotlinType): ConvertedType = when {
    type.isUnit -> ConvertedUnit()
    type.isInt -> ConvertedInt()
    else -> throw NotImplementedError()
}
```

Reply:
Done